### PR TITLE
Use URIs throughout the Language Server requests and responses.

### LINKED ISSUE #78: Fix conversion from URI to Path for all uris in requests

### LINKED ISSUE #79: Convert all response absolute paths to uris
Ex: Diagnostic uri

### PR DESCRIPTION
**Pull request by natgabb**
- Allows requests to use URIs relative to the workspace root and absolute paths.

Closes #78
Closes #79

### CODE REVIEW

File: groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    }
+
+    /**
+     * Returns the given absolutePath as a file URI.
+     * @throws IllegalArgumentException if absolutePath is not an absolute path
+     */
+    public static String absolutePathAsFileUri(Path absolutePath) {
+        checkArgument(absolutePath.isAbsolute(), "absolutePath must be absolute");
+        return "file:" + absolutePath.normalize().toString();
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java]: return Paths.get(string).toUri();

File: groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        } catch (Exception e) {
+            return Paths.get(uri).toAbsolutePath().normalize();
+        }
+    }
+
+    /**
+     * Normalizes the given URI and resolves it on the given absolutePathOfRoot if it is not already an absolute path.
+     * @throws IllegalArgumentException if absolutePathOfRoot is not an absolute path
+     */
+    public static Path resolveToRoot(Path absolutePathOfRoot, String uri) {
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java]: `absolutePathOfRoot` -> `absoluteRootPath`?

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java [hunk 3]
```diff
@@ -355,18 +353,18 @@ private void resetCompilationUnit() {
     }
 
     private void parseAllSymbols() {
-        Map<String, Set<SymbolInformation>> newFileSymbols = Maps.newHashMap();
+        Map<Path, Set<SymbolInformation>> newFileSymbols = Maps.newHashMap();
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java]: why is the key here a path not uri? haven't finished reading all your code yet but won't that save some conversion?

File: groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java (hunk 1 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java]: :o ! I'm going to remove this method altogether

File: groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java (hunk 2 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/util/Uris.java]: done

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java (hunk 3 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java]: yeah using a URI is a good idea! fixed

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyTextDocumentService.java [hunk 4]
```diff
@@ -208,13 +210,8 @@ private void publishDiagnostics(Set<Diagnostic> diagnostics) {
...
 
-    private Path resolveUriToWorkspaceRoot(String uri) {
-        Path filePath = Paths.get(uri);
-        return filePath.isAbsolute() ? filePath : provider.get().getWorkspaceRoot().resolve(uri).toAbsolutePath();
-    }
-
-    private void assertFileExists(Path uri) {
-        checkArgument(uri.toFile().exists(), String.format("Uri '%s' does not exist", uri));
+    private void assertFileExists(URI uri) {
+        checkArgument(Paths.get(uri).toFile().exists(), String.format("Uri '%s' does not exist", uri));
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyTextDocumentService.java]: checkArgument(new File(uri).exists(), String.format("Uri '%s' does not exist", uri));
https://docs.oracle.com/javase/7/docs/api/java/io/File.html#File(java.net.URI)

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java [hunk 5]
```diff
@@ -401,18 +398,26 @@ private void parseAllSymbols() {
...
-            newFileSymbols.put(getWorkspaceUri(sourcePath), symbols);
+            newFileSymbols.put(getWorkspaceUri(sourceUri), symbols);
         });
         // Set the new references and new symbols
         typeReferences = newTypeReferences;
         fileSymbols = newFileSymbols;
     }
 
+    private URI getWorkspaceUri(URI uri) {
+        // In the case that it's already relative to the workspace, we still convert it into a Path and then back into a
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java:414]: Will making sure workspaceRoot always have three slashes work here?

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyTextDocumentService.java (hunk 4 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyTextDocumentService.java]: done

**Comment by jiahuijiang** [comment]
Looks good after one last comment!

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java (hunk 5 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovycWrapper.java:414]: Not sure I follow?

**Pull request merged**